### ISSUE TITLE
QA Dashboard: E2E Metrics

### ISSUE THREAD

**Issue by JoeTice** [department-of-veterans-affairs/va.gov-team#22332]
## Issue Description
As a VFS Team member, I'd like to understand how well the product I'm developing is scoring based on platform QA standards, specifically in the area of: E2E Testing.

---
## Tasks
- [x] Survey the industry to determine what other organizations track in these areas
- [x] Look into what metrics are available through our current set of tools
   - [x] Look into what other tools may be able to provide to gain additional insights
- [x] Discuss any additional metrics that we'd like to know and research those metrics to determine if it can be tracked
- [x] Identify which metrics we will track (supporting metrics)
   - [x] Identify any key metrics that will be used to score the product in this area (KPIs)

## Acceptance Criteria
- [x] We will have a final list of KPIs and supporting metrics
- [x] We have identify sources for each of those metrics
- [x] We have tested to ensure those metrics are accessible
---

## How to configure this issue
- [ ] **Attached to a Milestone** (when will this be completed?)
- [ ] **Attached to an Epic** (what body of work is this a part of?)
- [ ] **Labeled with Team** (`product support`, `analytics-insights`, `operations`, `service-design`, `tools-be`, `tools-fe`)
- [ ] **Labeled with Practice Area** (`backend`, `frontend`, `devops`, `design`, `research`, `product`, `ia`, `qa`, `analytics`, `contact center`, `research`, `accessibility`, `content`)
- [ ] **Labeled with Type** (`bug`, `request`, `discovery`, `documentation`, etc.)

**Comment by holdenhinkle** [comment]
From [Ticket #17228](https://app.zenhub.com/workspaces/vsp-5cedc9cce6e3335dc5a49fc4/issues/department-of-veterans-affairs/va.gov-team/17228):

# Cypress Test Report
Using the following Mochawesome packages:

- https://github.com/adamgruber/mochawesome
- https://github.com/Antontelesh/mochawesome-merge

We can create a useful, nice looking, interactive Cypress Test Results page:

![image.png](https://images.zenhubusercontent.com/5f85bcde4f4764585e8ff820/a6f64b1b-7750-4336-8c92-7962ae50c43d)

We can grab the `json` that Mochawesome generates to create reports like this - https://nottyo.github.io/cypress-e2e-code-coverage/cypress/report/mochawesome-report/mochawesome.html

Mochawesome generates a json file like the following for each spec file:
File: `cypress/report/mochawesome-report/mochawesome_04062021_104000.json`

```json
"{
  ""stats"": {
    ""suites"": 1,
    ""tests"": 4,
    ""passes"": 4,
    ""pending"": 0,
    ""failures"": 0,
    ""start"": ""2021-04-06T14:39:56.174Z"",
    ""end"": ""2021-04-06T14:40:00.989Z"",
    ""duration"": 4815,
    ""testsRegistered"": 4,
    ""passPercent"": 100,
    ""pendingPercent"": 0,
    ""other"": 0,
    ""hasOther"": false,
    ""skipped"": 0,
    ""hasSkipped"": false
  },
  ""results"": [
    {
      ""uuid"": ""703365e3-d8d8-4aad-88f7-564ded444c37"",
      ""title"": """",
      ""fullFile"": ""src/applications/disability-benefits/all-claims/tests/all-claims-wizard.cypress.spec.js"",
      ""file"": ""src/applications/disability-benefits/all-claims/tests/all-claims-wizard.cypress.spec.js"",
      ""beforeHooks"": [],
      ""afterHooks"": [],
      ""tests"": [],
      ""suites"": [
        {
          ""uuid"": ""b9995547-4d0a-4168-b79c-778ba49b2560"",
          ""title"": ""526 wizard"",
          ""fullFile"": """",
          ""file"": """",
          ""beforeHooks"": [],
          ""afterHooks"": [],
          ""tests"": [
            {
              ""title"": ""should show the form wizard"",
              ""fullTitle"": ""526 wizard should show the form wizard"",
              ""timedOut"": null,
              ""duration"": 1408,
              ""state"": ""passed"",
              ""speed"": ""slow"",
              ""pass"": true,
              ""fail"": false,
              ""pending"": false,
              ""context"": null,
              ""code"": ""cy.url().should('include', _constants__WEBPACK_IMPORTED_MODULE_2__[\""DISABILITY_526_V2_ROOT_URL\""]);\ncy.get('h1').should('have.text', 'File for disability compensation');\ncy.axeCheck();"",
              ""err"": {},
              ""uuid"": ""2bccc1b5-3b74-4d7f-b198-4be82778c083"",
              ""parentUUID"": ""b9995547-4d0a-4168-b79c-778ba49b2560"",
              ""isHook"": false,
              ""skipped"": false
            },
            {
              ""title"": ""should show disagreeing with a decision questions & alert"",
              ""fullTitle"": ""526 wizard should show disagreeing with a decision questions & alert"",
              ""timedOut"": null,
              ""duration"": 1122,
              ""state"": ""passed"",
              ""speed"": ""slow"",
              ""pass"": true,
              ""fail"": false,
              ""pending"": false,
              ""context"": null,
              ""code"": ""cy.get('[type=\""radio\""][value=\""appeals\""]').check(checkOpt);\ncy.checkFormChange({\n  label: 'Are you on active duty right now?',\n  value: 'no-appeals'\n});\ncy.get('[type=\""radio\""][value=\""disagree-file-claim\""]').check(checkOpt);\ncy.checkFormChange({\n  label: 'Are you filing a new claim or are you disagreeing with a VA decision on an earlier claim?',\n  value: 'disagreeing'\n});\ncy.checkFormAlert('disagree with VA decision, needs a decision review');\ncy.axeCheck();"",
              ""err"": {},
              ""uuid"": ""40491446-1fd7-492d-90ce-9f577c820d8c"",
              ""parentUUID"": ""b9995547-4d0a-4168-b79c-778ba49b2560"",
              ""isHook"": false,
              ""skipped"": false
            },
            {
              ""title"": ""should show BDD questions & start button"",
              ""fullTitle"": ""526 wizard should show BDD questions & start button"",
              ""timedOut"": null,
              ""duration"": 1292,
              ""state"": ""passed"",
              ""speed"": ""slow"",
              ""pass"": true,
              ""fail"": false,
              ""pending"": false,
              ""context"": null,
              ""code"": ""cy.get('[type=\""radio\""][value=\""bdd\""]').check(checkOpt);\ncy.checkFormChange({\n  label: 'Are you on active duty right now?',\n  value: 'yes-bdd'\n});\ncy.get('.form-datefield-month').should('exist');\ncy.findByLabelText(/month/i).select(mockMonth120);\ncy.findByLabelText(/day/i).select(mockDay120);\ncy.findByLabelText(/year/i).clear().type(mockYear120);\ncy.checkStorage(_constants__WEBPACK_IMPORTED_MODULE_2__[\""FORM_STATUS_BDD\""], 'true');\ncy.checkStorage(_constants__WEBPACK_IMPORTED_MODULE_2__[\""SAVED_SEPARATION_DATE\""], mockDate);\ncy.checkFormChange({\n  type: 'usa-date-of-birth',\n  label: 'What’s the date or anticipated date of your release from active duty?',\n  value: mockDate\n});\ncy.checkCallToAction();\ncy.axeCheck();"",
              ""err"": {},
              ""uuid"": ""808ae2ea-df40-48ff-ab59-efa50c0f414c"",
              ""parentUUID"": ""b9995547-4d0a-4168-b79c-778ba49b2560"",
              ""isHook"": false,
              ""skipped"": false
            },
            {
              ""title"": ""should start all-claims/original claims flow"",
              ""fullTitle"": ""526 wizard should start all-claims/original claims flow"",
              ""timedOut"": null,
              ""duration"": 970,
              ""state"": ""passed"",
              ""speed"": ""slow"",
              ""pass"": true,
              ""fail"": false,
              ""pending"": false,
              ""context"": null,
              ""code"": ""var h1Text = 'File for disability compensation'; // starts with focus on breadcrumb\ncy.focused().should('have.class', 'va-nav-breadcrumbs-list');\ncy.get('h1').should('have.text', h1Text);\ncy.get('[type=\""radio\""][value=\""appeals\""]').check(checkOpt);\ncy.checkFormChange({\n  label: 'Are you on active duty right now?',\n  value: 'no-appeals'\n});\ncy.get('[type=\""radio\""][value=\""file-claim\""]').check(checkOpt);\ncy.checkStorage(_constants__WEBPACK_IMPORTED_MODULE_2__[\""FORM_STATUS_BDD\""], undefined);\ncy.checkStorage(_constants__WEBPACK_IMPORTED_MODULE_2__[\""SAVED_SEPARATION_DATE\""], undefined); // learn more link\ncy.get('a[href*=\""disability/how-to-file-claim/\""]').should('exist');\ncy.checkFormChange({\n  label: 'Are you filing a new claim or are you disagreeing with a VA decision on an earlier claim?',\n  value: 'new-worse'\n});\ncy.axeCheck(); // start form\nvar h1Addition = ' with VA Form 21-526EZ';\ncy.findAllByText(/disability claim online/i, {\n  selector: 'a'\n}).first().click(); // title changes & gets focus\ncy.get('h1').should('have.text', h1Text + h1Addition);\ncy.focused().should('have.text', h1Text + h1Addition);\ncy.checkStorage(_constants__WEBPACK_IMPORTED_MODULE_2__[\""WIZARD_STATUS\""], 'complete');\ncy.axeCheck();"",
              ""err"": {},
              ""uuid"": ""c1d385cb-be8e-40d9-8efb-b2b099c68468"",
              ""parentUUID"": ""b9995547-4d0a-4168-b79c-778ba49b2560"",
              ""isHook"": false,
              ""skipped"": false
            }
          ],
          ""suites"": [],
          ""passes"": [
            ""2bccc1b5-3b74-4d7f-b198-4be82778c083"",
            ""40491446-1fd7-492d-90ce-9f577c820d8c"",
            ""808ae2ea-df40-48ff-ab59-efa50c0f414c"",
            ""c1d385cb-be8e-40d9-8efb-b2b099c68468""
          ],
          ""failures"": [],
          ""pending"": [],
          ""skipped"": [],
          ""duration"": 4792,
          ""root"": false,
          ""rootEmpty"": false,
          ""_timeout"": 2000
        }
      ],
      ""passes"": [],
      ""failures"": [],
      ""pending"": [],
      ""skipped"": [],
      ""duration"": 0,
      ""root"": true,
      ""rootEmpty"": true,
      ""_timeout"": 2000
    }
  ],
  ""meta"": {
    ""mocha"": {
      ""version"": ""7.0.1""
    },
    ""mochawesome"": {
      ""options"": {
        ""quiet"": false,
        ""reportFilename"": ""mochawesome"",
        ""saveHtml"": false,
        ""saveJson"": true,
        ""consoleReporter"": ""spec"",
        ""useInlineDiffs"": false,
        ""code"": true
      },
      ""version"": ""6.2.2""
    },
    ""marge"": {
      ""options"": {
        ""reportDir"": ""cypress/report/mochawesome-report"",
        ""overwrite"": false,
        ""html"": false,
        ""json"": true,
        ""timestamp"": ""mmddyyyy_HHMMss""
      },
      ""version"": ""5.2.0""
    }
  }
}"
```

After all tests are executed, it merges the test results in `mochawesome.json`.

PR: https://github.com/department-of-veterans-affairs/vets-website/pull/16655
Run command: `node cypress_runner`

**Comment by holdenhinkle** [comment]
Working from this repo again - https://github.com/department-of-veterans-affairs/vets-website/pull/16655

- I fixed the paths to failed test images
- I updated `cy:run` to run the custom Cypress test runner that generates the Mochawesome Report for Cypress test executions

I also met with Peter to discuss how we want to handle the deeply nested data in the `mochawesome.json` file. The `json` file has three 'top-level' objects:

1. `stats`, which contains a summary of the test execution results (totals for pass, skipped, failed, etc.).
2. `results`, which contains and array of test objects, with each object corresponding to the tests in a `spec` file. The data in each object is deeply nested, matching the nested structure of the `describe` and `it` blocks in each `spec` file.
3. `meta`, which contains version and configuration information for Mocha, Mochawesome, and Marge.

We're going to create a BigQuery table that holds the `stats` data, with an `id` column ( with a `UUID`). The table will be called something like `cypress_test_suite_executions`.

Then we're going to create a table to hold each `results` object, with a `cypress_test_suite_executions_id` column, to reference the test suite execution to which the record belongs. We're going to make use of BigQuery's `record` type and `repeatable` fields to hold the nested test results data.

I'm inclined to not store the `meta` data.

Peter and I spoke about how we might link each result to a Mochawesome report item. Each item in the report includes an `id` so it's easy to link.

We talked about how we might aggregate the test results from each Mochawesome report that's generated from running the Cypress tests in parallel. I found [this article](https://tsh.io/blog/cypress-and-mochawesome-reporting/#script-configuration-3) which describes exactly how to do this using the `mochawesome-merge` package.

**Comment by holdenhinkle** [comment]
CI steps for creating a Cypress Mochawesome Report and posting the Cypress test results to BigQuery:

### Step 1
Cypress tests are divided into n steps, each step running an arbitrary set of Cypress tests in a Docker container, each container running in parallel. While Cypress tests are being executed in each container, Mochawesome will generate a `json` results file for each spec. After the Cypress tests have completed, the `json` files will be copied to the host into a the `cypress/report/mochawesome-report` directory. 

### Step 2
This CI step is dependent on Step 1.

This step will merge the `json` files (creating a `mochawesome.json` file), generate the Mochawesome report, rename the report from `mochawesome.html` to something like `2021-05-26T12:29:50.898Z.html` (a date-time string) and post the report to a GitHub repo for serving the static Mochawesome reports. The directory structure will look like this:

```
mochawesome-reports
  assets
  2021-05-26T12:29:50.898Z.html
  2021-05-26T12:11:50.898Z.html
  2021-04-26T11:15:50.898Z.html
```

### Step 3
This CI step is dependent on Step 2  because we need the name of the Mochawesome report that was generated. This step will parse the `mochawesome.json` file into row objects that will be inserted in two BigQuery tables. While parsing the `json` file, a field called `mochawesome_report_link` will be created that will be set to the url of the new report in the GitHub static site.

**Comment by holdenhinkle** [comment]
Peter and I talked about the ideas in my last post (see above). We agreed it would be better to post the video of failing tests in the Mochawesome Reports instead of the image.

Peter isn't sure we can copy the `json` files from each container running the Cypress tests in parallel back to the main build and I'm not familiar with how this part of GHA works yet. It's seems we can store the `json` files as artifacts and make them available to another GHA where can merge them and generate the Mochawesome report. We'd then post the report and the failed videos to the GitHub pages repo where the reports will be served.

Questions:
- Should we store the videos in GH or in an S3 bucket?
- Should we delete videos that are older than, say, 30 days, each day?

Peter found this post which describes hows to generate a Mochawesome report in GitHub Actions and host it in GitHub pages, which mirrors my idea--[Publish your Cypress Test Report with GitHub Actions](https://medium.com/swlh/publish-your-cypress-test-report-with-github-actions-47248788713a).

It would be great if we can link to or embed the Mochawesome report in an iFrame in the Checks tab in GH for each build.

Immediate next step:
- Parse the Mochawesome `json` file and post the data to the appropriate BigQuery tables.

**Comment by holdenhinkle** [comment]
Quick update. I need to create a new ticket to track the following:

I created a [new branch/draft pr on the Testing Tools Team Dashboard Data repo](https://github.com/department-of-veterans-affairs/testing-tools-team-dashboard-data/pull/17) to:

- merge Mochawesome json test rests json file (a results json file is created for each Cypress spec file)
- generate a Mochawesome report
- create the rows for the following BigQuery tables - `cypress_test_suite_executions`, `cypress_tests`
- create the schema for those tables
- posted the row data to the tables

All that's complete.

Here are the tables with data in BigQuery:

cypress_test_suite_executions - https://console.cloud.google.com/bigquery?project=vsp-analytics-and-insights&ws=!1m10!1m4!4m3!1svsp-analytics-and-insights!2svsp_testing_tools!3scypress_test_results!1m4!4m3!1svsp-analytics-and-insights!2svsp_testing_tools!3scypress_test_suite_executions&page=table&d=vsp_testing_tools&p=vsp-analytics-and-insights&t=cypress_test_suite_executions

cypress_test_results table - https://console.cloud.google.com/bigquery?project=vsp-analytics-and-insights&ws=!1m10!1m4!4m3!1svsp-analytics-and-insights!2svsp_testing_tools!3scypress_test_results!1m4!4m3!1svsp-analytics-and-insights!2svsp_testing_tools!3scypress_test_suite_executions&page=table&d=vsp_testing_tools&p=vsp-analytics-and-insights&t=cypress_test_results

- I created a repo that we'll post the test reports to - https://github.com/department-of-veterans-affairs/testing-reports
- I activated GitHub pages. GitHub will serve the reports.
- I created a `cypress-tests` directory, loaded the Mochawesome assets (which don't change), and loaded a sample Mochawesome report
- Mochawesome report - https://reimagined-telegram-c11df05a.pages.github.io/cypress-reports/483f47e3-1c6a-48a4-a6bc-30906485ef7f.html

(I still need to update the report generator to include videos of failing tests, etc.)

I'll continue tomorrow!

**Comment by holdenhinkle** [comment]
This is done.

**Comment by pjhill** [comment]
Huzzah, this is indeed done. Post-MVP changes can be worked under a new ticket.
